### PR TITLE
[FLINK-2954] Add config parameter for passing environment variables to YARN

### PR DESCRIPTION
**Pull request by rmetzger**

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java [hunk 1]
```diff
@@ -243,6 +243,20 @@
...
+	 * 	yarn.application-master.env.LD_LIBRARY_PATH: "/usr/lib/native"
+	 * in the flink-conf.yaml.
+	 */
+	public static final String YARN_APPLICATION_MASTER_ENV_PREFIX = "yarn.application-master.env.";
+
+	/**
+	 * Similar to the {@see YARN_APPLICATION_MASTER_ENV_PREFIX}, this configuration prefix allows
+	 * setting custom environment variables.
+	 */
+	public static final String YARN_TASK_MANAGER_ENV_PREFIX = "yarn.taskmanager.env.";
```
Comment (mxm) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java]: I wonder about the naming here. Maybe this should be `YARN_RESOURCE_MANAGER_ENV_PREFIX`? Or change `YARN_APPLICATION_MASTER_ENV_PREFIX` to `YARN_JOB_MANAGER_ENV_PREFIX`?

File: flink-yarn-tests/src/main/java/org/apache/flink/yarn/UtilsTest.java [hunk 2]
```diff
@@ -97,6 +98,18 @@ public void tooMuchCutoff() {
...
+		Configuration testConf = new Configuration();
+		testConf.setString("yarn.application-master.env.LD_LIBRARY_PATH", "/usr/lib/native");
+
+		Map<String, String> res = Utils.getEnvironmentVariables("yarn.application-master.env.", testConf);
+
+		Assert.assertEquals(1, res.size());
+		Map.Entry<String, String> entry = res.entrySet().iterator().next();
+		Assert.assertEquals("LD_LIBRARY_PATH", entry.getKey());
+		Assert.assertEquals("/usr/lib/native", entry.getValue());
+	}
```
Comment (mxm) [inline flink-yarn-tests/src/main/java/org/apache/flink/yarn/UtilsTest.java]: There is no test for the task manager variables.

File: flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java [hunk 3]
```diff
@@ -221,4 +210,22 @@ public static void addToEnvironment(Map<String, String> environment,
...
+	 *
+	 * @param envPrefix Prefix for the environment variables key
+	 * @param flinkConfiguration The Flink config to get the environment variable defintion from
+	 */
+	public static Map<String, String> getEnvironmentVariables(String envPrefix, org.apache.flink.configuration.Configuration flinkConfiguration) {
+		Map<String, String> result  = new HashMap<>();
+		for(Map.Entry<String, String> entry: flinkConfiguration.toMap().entrySet()) {
+			if(entry.getKey().startsWith(envPrefix)) {
+				// remove prefix
+				String key = entry.getKey().substring(envPrefix.length());
```
Comment (mxm) [inline flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java]: What happens if the key is `envPrefix`? There is little validation here.

**Comment by mxm** [comment]
Looks good except for some minor issues.

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java (hunk 1 above)
Comment (rmetzger) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java]: I think `YARN_RESOURCE_MANAGER_ENV_PREFIX` is just wrong bc its a service of YARN, not of Flink (http://hortonworks.com/blog/apache-hadoop-yarn-resourcemanager/).

Whether to use AM or JM is a good question. In the YARN context, AM and JM are basically the same. I think AM is a bit more consistent.

File: flink-yarn-tests/src/main/java/org/apache/flink/yarn/UtilsTest.java (hunk 2 above)
Comment (rmetzger) [inline flink-yarn-tests/src/main/java/org/apache/flink/yarn/UtilsTest.java]: I think its okay to assume that the method will work for all prefix strings.
This test is only about the method for extracting the variables.

File: flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java (hunk 3 above)
Comment (rmetzger) [inline flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java]: I agree. I'll harden the method.

**Comment by rmetzger** [comment]
Thank you for the good review. I addressed some of your concerns.

**Comment by rmetzger** [comment]
Thanks for the review. I'll merge it soon

**Pull request merged**